File: C/timing_leak_example.cpp
#include <iostream>
#include <chrono>

using namespace std;
using namespace std::chrono;

// Insecure function (leaks timing information)
bool insecureCheck(const string &input, const string &correct) {
    for (size_t i = 0; i < correct.length(); i++) {
        if (input[i] != correct[i]) {
            return false; // Early exit leads to timing leak
        }
    }
    return true;
}

// Secure function (constant-time comparison)
bool secureCheck(const string &input, const string &correct) {
    int diff = 0;
    size_t len = correct.length();
    
    for (size_t i = 0; i < len; i++) {
        diff |= input[i] ^ correct[i];  // No branching
    }
    
    return diff == 0;
}

int main() {
    string correctPassword = "secure123";
    string input = "secure123";

    // Measure time for insecure check
    auto start = high_resolution_clock::now();
    bool result = insecureCheck(input, correctPassword);
    auto end = high_resolution_clock::now();
    auto duration = duration_cast<nanoseconds>(end - start);
    cout << "Insecure Check Time: " << duration.count() << " ns, Result: " << (result ? "Match" : "No Match") << endl;

    // Measure time for secure check
    start = high_resolution_clock::now();
    result = secureCheck(input, correctPassword);
    end = high_resolution_clock::now();
    duration = duration_cast<nanoseconds>(end - start);
    cout << "Secure Check Time: " << duration.count() << " ns, Result: " << (result ? "Match" : "No Match") << endl;

    return 0;
}

File: PHP/timing_leak_example.php
<?php

function insecureCheck($input, $correct) {
    // Bad case: Leaks timing due to early exit
    for ($i = 0; $i < strlen($correct); $i++) {
        if ($input[$i] !== $correct[$i]) {
            return false; // Timing leak occurs here
        }
    }
    return true;
}

function secureCheck($input, $correct) {
    // Good case: Constant-time comparison
    $diff = 0;
    $len = strlen($correct);
    
    for ($i = 0; $i < $len; $i++) {
        $diff |= ord($input[$i]) ^ ord($correct[$i]);
    }
    
    return $diff === 0;
}

$correctPassword = "secure123";
$input = "secure123";

// Insecure check
$start = microtime(true);
$result = insecureCheck($input, $correctPassword);
$end = microtime(true);
echo "Insecure Check Time: " . (($end - $start) * 1e6) . " µs, Result: " . ($result ? "Match" : "No Match") . "\\n";

// Secure check
$start = microtime(true);
$result = secureCheck($input, $correctPassword);
$end = microtime(true);
echo "Secure Check Time: " . (($end - $start) * 1e6) . " µs, Result: " . ($result ? "Match" : "No Match") . "\\n";

?>
